### PR TITLE
Add AuthService for for deploy-key authenticated git operations

### LINKED ISSUE #577: Refactor utils.SanitizeRepoURL function to not assume github and support https
`SanitizeRepoURL` currently looks like this:

```golang
func SanitizeRepoUrl(url string) string {
	trimmed := ""

	if !strings.HasSuffix(url, ".git") {
		url = url + ".git"
	}

	sshPrefix := "git@github.com:"
	if strings.HasPrefix(url, sshPrefix) {
		trimmed = strings.TrimPrefix(url, sshPrefix)
	}

	httpsPrefix := "https://github.com/"
	if strings.HasPrefix(url, httpsPrefix) {
		trimmed = strings.TrimPrefix(url, httpsPrefix)
	}

	if trimmed != "" {
		return "ssh://git@github.com/" + trimmed
	}

	return url
}
```

^^ This assumes Github and overwrite HTTPS urls. We will need to support any git provider. 

We also should consider adding some type safety to this. Passing strings around makes the code very hard to understand.

### LINKED ISSUE #579: Use deploy key for git operations
Follow up from #569

As of #569, we can get a temporary access token to do Git Provider operations. We use this token to upload a deploy-key so that we can do `git` operations, but currently only the Flux controllers utilize it for `git` operations.

We should use the deploy key we uploaded to do `git` operations to avoid utilizing the user's local environment to retrieve ssh keys. This will be crucial for running `wego` in the cluster as an API, where we do not have access to an `ssh-agent` or other mechanisms to get access to a key.

### PR DESCRIPTION
**Pull request by jpellizzari**
Closes #579
Closes #577 

Adds an `AuthService` object that can be used to handle creating or retrieving repository deploy keys. This allows us to create a git client decorated with a public key in memory without having to read the users local file system.

This will be crucial for doing operations in environments where we do not have access to a user environment, such as running as a pod on the cluster.

Does **NOT** currently handle the external config repo scenario. That case will still require keys from the user's environment. Followup created here: https://github.com/weaveworks/weave-gitops/issues/592

Tested manually and with unit tests. More acceptance test coverage will be added here: #581

### CODE REVIEW

File: pkg/kube/kubehttp.go
Comment (J-Thompson12) [inline pkg/kube/kubehttp.go:61]: I am curious about this name change. What makes it a raw client? When its used elsewhere its less clear that its a kube client now.

Comment (jpellizzari) [reply pkg/kube/kubehttp.go:61]: Only to further disambiguate it. It could have stayed the same given that the name of this variable is scoped to this function. I didn't want there to be any ambiguity for the reader of this function.

**Review by J-Thompson12 — APPROVED** [review]

File: pkg/services/auth/auth_test.go
Comment (palemtnrider) [inline pkg/services/auth/auth_test.go:86]: Test should verify that no deploy keys exist (based on the `It` clause. Without this check we might be reusing an environment and not actually testing the create.

File: pkg/utils/common.go
Comment (palemtnrider) [inline pkg/utils/common.go:121]: nit - both arguments are string so you can change the function to take (targetName, repoURL string)

Comment (palemtnrider) [reply pkg/utils/common.go:121]: Since this is an exported utility function, we should have a function comment.

File: pkg/utils/common.go
Comment (palemtnrider) [inline pkg/utils/common.go:125]: Should have go doc for the exported function

File: pkg/utils/common.go
Comment (palemtnrider) [inline pkg/utils/common.go:155]: Since this function is about starting a k8s test environment, we should fail if we can't start the manager instead of just printing a message.

**Review by palemtnrider — COMMENTED** [review]
Good to see the deploy key work being injected into the code base.  Some comments/questions/changes before I can approve.  Don't hesitate to reach out if something isn't clear or more context is necessary.

File: pkg/kube/kubehttp.go
Comment (jpellizzari) [reply pkg/kube/kubehttp.go:40]: The `client.Client` utilized by the `AuthService`

Comment (jpellizzari) [reply pkg/kube/kubehttp.go:40]: >  I'd leave it as part of the structure and remove it as an individual argument.

Sorry I originally misread. The `kube.Kube` implementation does not always have the `client.Client` as a dependency, only the `KubeHTTP` struct has it internally.

The other issue is that `client.Client` makes the entire `kube.Kube` interface unnecessary, so there is probably a bigger change that needs to be made around this interface.

File: test/acceptance/test/add_tests.go
Comment (jpellizzari) [reply test/acceptance/test/add_tests.go:10]: I did not create this file in this PR?

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:190]: It's an internal function, I think the name is OK as is.

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:199]: It's an internal function, I think the name is OK as is.

File: pkg/kube/kube.go
Comment (jpellizzari) [reply pkg/kube/kube.go:217]: Added to the `kubectl` command

File: .github/workflows/test.yml
Comment (JamWils) [reply .github/workflows/test.yml:107]: I agree, we can control the upgrade patch and know when it is the version change that is causing failures.

File: pkg/utils/common.go
Comment (jpellizzari) [reply pkg/utils/common.go]: Fixed. Calling the `Expect` func in the test helper

File: api/v1alpha1/application_types.go
Comment (jpellizzari) [reply api/v1alpha1/application_types.go:69]: https://app.zenhub.com/workspaces/weaveworks-foss-603cb6b1d4dff9000e242dc4/issues/weaveworks/weave-gitops/612

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go]: Added a `NormalizedRepoURL` type here: https://github.com/weaveworks/weave-gitops/pull/587/commits/c873e7dd4ae26102e681b386a7d10c8639d50f62

Consuming it here: https://github.com/weaveworks/weave-gitops/pull/587/commits/e0864b51bd9cacefc4b9a3d462de693cfa7e9da8

File: pkg/services/auth/auth.go
Comment (JamWils) [inline pkg/services/auth/auth.go:47]: I would make an interface for this and return that from this function instead of a pointer to the struct.  Albeit, the underlying object is still a pointer to a struct.  It's a pattern that need to start following across the codebase.

```
type AuthSomething interface {
	SetupDeployKey(ctx context.Context, name types.NamespacedName, targetName string, repo gitproviders.NormalizedRepoURL) error
	GitClient() (git.Git, error)
	GitProvider() gitproviders.GitProvider
}
```

File: pkg/services/auth/auth.go
Comment (JamWils) [inline pkg/services/auth/auth.go:98]: Shadow declaration on `err`.  Consider renaming to `errDeployKey` or something along those lines.

File: pkg/services/auth/auth.go
Comment (JamWils) [inline pkg/services/auth/auth.go:111]: Note: if you fix the above shadow variable then this will become a new shadow variable, so I would rename this one as well.

File: pkg/services/auth/auth.go
Comment (JamWils) [reply pkg/services/auth/auth.go:65]: Does it make sense to just return the gitClient from here instead of having a separate "getter" for the git client?  It would simplify the consumption of this implementation.  Looking at whee we use `SetupDeployKey` it is always called before `GitClient`.
The following code can be simplified from:
```
if err := authsvc.SetupDeployKey(ctx, name, targetName, normalizedUrl); err != nil {
    return fmt.Errorf("error setting up deploy keys: %w", err)
}

gitClient, err = authsvc.GitClient()
if err != nil {
    return fmt.Errorf("error getting authenticated git client: %w", err)
}
```
to:
```
gitClient, err = authsvc.SetupGitClient(ctx, name, targetName, normalizedUrl)
if err != nil {
    return fmt.Errorf("error setting up git client: %w", err)
}
```

Comment (JamWils) [reply pkg/services/auth/auth.go:65]: This could also eliminate the need for `ErrNoDeployKeysSetup`

File: pkg/services/auth/auth.go
Comment (JamWils) [inline pkg/services/auth/auth.go:203]: How do you feel about refactoring the fluxClient to return the `corev1.Secret` instead of `[]byte` from `CreateSecretGit`?  It is called in two places, here and `Add`.  In both use-cases it is unmarshaled immediately into the `*corev1.Secret`.  This will clean up the code a lot and will help set a precedent to move away from `[]byte`.

File: pkg/services/auth/auth.go
Comment (JamWils) [inline pkg/services/auth/auth.go:201]: I really advise returning just the struct `corev1.Secret` instead of the pointer.  I can tell you from experience that copying is cheap in Go, it will make the code simpler and easier to reason about over the long term as well.

File: pkg/utils/common.go
Comment (JamWils) [inline pkg/utils/common.go:141]: Can we put this in a different package like `testutils`?  It will be easier to catch if someone uses this in a place they shouldn't versus being mixed in with other `utils` functions that can be used within the app.

File: cmd/wego/app/add/cmd.go
Comment (JamWils) [inline cmd/wego/app/add/cmd.go:105]: The isHelmChart could be a function on the Params struct itself.

Just thinking out loud here and this isn't really applicable to this PR.  However, it would be nice if `params` used embedded structs and we could return either one depending on whether we are dealing with `helm` or a `repo`.  This would simplify the code over the long-term.  Again, just a random thought outside of the scope of this PR.

File: test/acceptance/test/add_tests.go
Comment (JamWils) [reply test/acceptance/test/add_tests.go:10]: All of the files end with `tests`, we could open a separate issue to update all of them if we feel that strongly about it.  I wouldn't just change the one and since it is more than one I wouldn't do it in this PR.

File: cmd/wego/app/add/cmd.go
Comment (JamWils) [inline cmd/wego/app/add/cmd.go:152]: This function has numerous shadow variables for `err`.  Consider renaming.

File: pkg/services/auth/auth.go
Comment (JamWils) [inline pkg/services/auth/auth.go:159]: This isn't used, consider removing.

**Review by JamWils — CHANGES_REQUESTED** [review]
I'm getting cross-eyed on this PR and need to submit the current comments.  Overall, the PR looks good.  I would like to double check some business logic on a final pass.

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:98]: `err` has to be nil before this line. Do we need to avoid shadowing `err` everywhere?

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:203]: That would be great, but out of scope for this PR. Lots of other code will have to change.

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:201]: I'm following how the official k8s client libs do it: https://pkg.go.dev/k8s.io/client-go@v0.22.0/kubernetes/typed/apps/v1#DeploymentInterface

You are going to have to pry my "always pass by reference" JavaScript habits from my cold, dead fingers :wink:. 

Let's come up with some sort of guide on this for the team?

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:47]: I had thought about an `interface` but didn't anticipate this needing to be mocked, since it provides an existing "client" `interface` in `.GitClient()`.

I think I will refactor back to an `interface`, since some future tests might want to mock it.

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go]: This behavior has now been combined into a single function and the `GitClient()` and `GitProvider()` methods have been removed.

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:65]: Refactored to return the `.GitClient()`

File: pkg/utils/common.go
Comment (jpellizzari) [reply pkg/utils/common.go]: Moved

File: pkg/gitproviders/provider_test.go
Comment (jpellizzari) [reply pkg/gitproviders/provider_test.go:766]: I went through and tried this, but it won't work in this case, since these are function calls.

File: cmd/wego/app/add/cmd.go
Comment (JamWils) [reply cmd/wego/app/add/cmd.go:130]: That sounds like a plan.  I tried holding back in a lot of other areas where there were comments.  I'll bring it up at retro on Friday.

File: pkg/services/auth/auth.go
Comment (JamWils) [reply pkg/services/auth/auth.go:94]: At my last place we always avoided shadow variables.  Eventually it always would bite us and it was always subtle and took a lot of debugging to find out that was the issue.  After enough times it was a hard no.  You don't have to fix them, but one of them somewhere will get us one day.

File: pkg/services/auth/auth.go
Comment (JamWils) [reply pkg/services/auth/auth.go:177]: Bah, I missed in the other location that the secret is then passed to kube.Apply which takes `[]byte`.  Yeah, pull that thread and it is bigger than I thought it would be.

File: pkg/services/app/app.go
Comment (JamWils) [reply pkg/services/app/app.go:196]: Yeah, I was more referring to the switch statement below that is returning the errors.  I get testing the happy path but it is also important to test the sad path and make sure that is captured.  It lowers the friction to being able to refactor code.

File: pkg/services/auth/auth.go
Comment (JamWils) [reply pkg/services/auth/auth.go:94]: > err has to be nil before this line. Do we need to avoid shadowing err everywhere?

Also, very valid point.

File: pkg/services/auth/auth.go
Comment (JamWils) [reply pkg/services/auth/auth.go]: My last place of employment did the same to me.  I'm just passing on the habit :)

We were allocating a ridiculous number of structs and learned the hard way that pointers on the heap make machines sad if you allocate enough of them.  I get we won't meet the volume that we experienced.  I also find it easier to reason about.  Anyway, your reasoning for following the k8s standards is extremely valid as well.

File: pkg/kube/kube.go
Comment (JamWils) [reply pkg/kube/kube.go:67]: > I was trying to keep consistent with that/those lib

Hard to aruge with that, if we create a time series database then I'll raise the issue then :)

File: pkg/gitproviders/provider.go
Comment (JamWils) [reply pkg/gitproviders/provider.go:575]: I do like to test the sad path in my code as well.  I find it rather easy in Go compared to other languages.  I'll leave it up to you though.

File: cmd/wego/app/add/cmd.go
Comment (JamWils) [reply cmd/wego/app/add/cmd.go:151]: Code evolves.  It was more just the random check on `isHelmChart` throughout the function.

File: cmd/wego/app/add/cmd.go
Comment (JamWils) [reply cmd/wego/app/add/cmd.go:130]: I'm all about being able to just read code :).   I'll find a time to bring it up with the team.

File: cmd/wego/app/add/cmd.go
Comment (JamWils) [reply cmd/wego/app/add/cmd.go:152]: I'm going to resolve this, but leave it to you to decide what you want to do.  Again, they are subtle when they appear.

File: pkg/services/auth/auth.go
Comment (JamWils) [reply pkg/services/auth/auth.go:42]: I'll resolve this, but I'm telling you the variable name `k8sClient` tells me all I need to know :)

**Review by JamWils — APPROVED** [review]

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:42]: We don't use the `client.Client` anywhere else directly, so I wanted to save someone a "WTF" moment when they look at this. Speaks to the bigger question about comments I guess.

File: pkg/kube/kubehttp.go
Comment (jrryjcksn) [reply pkg/kube/kubehttp.go:40]: In fact, given the relatively small number of calls we make to `client.Client`, we could just have the `KubeHTTP` struct implement an interface that extends `kube.Kube` and not expose the raw client at all. That would let us mock those calls if we want to as well so we wouldn't always have to spin up a k8s server. Also, it would make our code more approachable to a potential open source contributor. They would know exactly what we do with Kubernetes. If we pass a raw client around, they have to assume we might do literally anything supported by Kubernetes.

File: cmd/wego/app/add/cmd.go
Comment (jpellizzari) [reply cmd/wego/app/add/cmd.go:115]: I think that's do-able; I would like to try that on whichever command consumes this next so the function matches exactly what we need.

File: cmd/wego/app/add/cmd.go
Comment (jpellizzari) [reply cmd/wego/app/add/cmd.go:105]: Good point :+1:

File: pkg/kube/kube.go
Comment (jpellizzari) [reply pkg/kube/kube.go:254]: Probably to avoid a cyclic import, and `kube` made as much sense as anywhere else at the time. I will look at moving it there.

File: pkg/kube/kubehttp.go
Comment (jpellizzari) [reply pkg/kube/kubehttp.go:40]: I would say that `client.Client` is just as easily mock-able (with some code generation), and gives us the added option of using the `envtest` package for more integration-y, non-mocked tests. T[he interfaces are also broken up by reader/writer if we want to get more specific](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#Client).

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go:99]: @jrryjcksn I think we need to delete the deploy key on the Provider, create a new one, and push it/save it on cluster. WDYT?

Comment (jrryjcksn) [reply pkg/services/auth/auth.go:99]: Agreed. :+1:

File: pkg/kube/kubehttp.go
Comment (jpellizzari) [reply pkg/kube/kubehttp.go:40]: Closing for now. I assume we will discuss further before changing that interface.

File: cmd/wego/app/add/cmd.go
Comment (jpellizzari) [reply cmd/wego/app/add/cmd.go:115]: Leaving as is for now and marking as resolved.

File: pkg/services/auth/auth.go
Comment (jpellizzari) [reply pkg/services/auth/auth.go]: @jrryjcksn I changed this to create a new key and re-upload, but NOT remove the old key. Doing destructive operations felt like a bad move. I am assuming the Git Provider will allow for the same key name so long as the SHAs are different.

**Review by jrryjcksn — APPROVED** [review]
LGTM

**Pull request merged**
